### ISSUE TITLE
Handshake middleware authentication

### ISSUE THREAD

**Issue by JasonPaulos** [TopCloud/socketcluster#53]
I'm currently developing an application that makes use of the handshake middleware to authenticate my users when they open up a socket connection. I'm using the module passport.socketcluster (https://github.com/artema/passport.socketcluster) to help out, and here's a sample of the code I'm using:
```
sc.addMiddleware(sc.MIDDLEWARE_HANDSHAKE, passportSocketCluster.handshake({
	cookieParser: cookieParser,
	key: cookieKey,
	secret: cookieSecret,
	store: store,
	passport: passport,
	callback: function(err, req, user, next){
		if(!err){
			req.socket.setAuthToken({ //authenticates socket only if user information was loaded
				profile: {
					name: user._json.personaname,
					avatar: user._json.avatar
				}
			});
		}
		next();
	}
}));
//Later on...
sc.on('connection', function(socket){
	var authToken = socket.getAuthToken();
	
	socket.on('state request', function(data){
		var state = {};
		
		if(authToken){
			state.profile = authToken.profile;
		}
		
		socket.emit('state', state);
	});
});
```
Basically, if a user has logged in, the server sends a 'state' event with profile information, but if the user hasn't logged in, then profile information isn't sent. Everything works when the user hasn't logged in, but when they have, the authToken is still null.

I've had a hard time finding documentation about the req parameter for the middleware handshake, and I was wondering if req.socket is a reference to the actual socket and if how I'm using req.socket.setAuthToken(...) is correct. If not, what are my other options for passing information from the handshake middleware to sockets now that socket.session is no longer available? Any advice will be appreciated.

**Comment by jondubois** [comment]
I don't know if passport.socketcluster will work with the latest version of SC. I should probably get in touch with the author to check. You could make a pull request if it's not :)

The req object in the handshake middleware is a regular Node.js HTTP request object https://nodejs.org/api/http.html#http_http_incomingmessage.

The handshake middleware function gets called just before the WebSocket is created, so the WebSocket doesn't exist yet at that point. You can use the req object to do stuff like read headers, cookies (auth tokens) etc... to decide whether or not to allow the client to establish a persistent WebSocket (or polling) connection with the server. This lets you do stuff like allow the user to connect over WebSocket only after they have logged in with HTTP first... (It prevents unauthorized clients from hogging up connections on your server).

...You may not necessarily need this feature; another approach would be to allow any client to create a websocket connection with your server (without special authentication) but then you do the access control at the 'publish', 'emit', 'subscribe' stages (using the appropriate middleware).

With the second approach, you effectively let anyone connect to your server, but if they want to interact with channels or send events, then they will need to have the right auth token to do that.

Note that there have been significant changes in how auth works in SC since v1.3.0. Now it's following the JWT (token) specification (https://tools.ietf.org/html/draft-ietf-oauth-json-web-token-32) which is the foundation for OAuth (instead of SocketCluster sessions). The new JWT engine should play nice with third-party OAuth providers like Facebook... although I haven't yet tried this feature yet ;(

So yeah, in summary, you don't have to use the handshake middleware if you don't want to unless you are really worried about denial of service attacks and you're too lazy to monitor your system :p

**Comment by jondubois** [comment]
@JasonPaulos Also, the req.socket object refers to the TCP socket (not WebSocket).

The purpose of the auth token in SC is to keep the user logged in if their connection crashes or if they revisit the page at a later time - setAuthToken() is something you do **after** the login has taken place (successfully) by that point your user should already have access to your WebSocket connection.

**Comment by JasonPaulos** [comment]
Thanks for informing me of the details of the req parameter and how authentication works in the current version of SC. I have to agree that auth tokens do seem to be a more elegant solution than sessions, but I had planned my current project around being able to "soft authenticate" sockets with their passport session if they had already logged in. And, unfortunately, sending login credentials though sockets isn't an option with the passport strategy I'm using (https://github.com/liamcurry/passport-steam). With passport-steam, login credentials are sent through Steam and then profile information is sent to my app. This information is then stored in sessions, so associating a session with a socket is a key step in my current workflow.

To accommodate this necessity, I see no other option than to modify my SC source slightly. I've had to add the line ```this.profile = req.profile;``` to the constructor of SCSocket here: https://github.com/TopCloud/socketcluster/blob/master/lib/server/scsocket.js#L65. And in the passportSocketCluster.handshake callback function in my own app, I've set req.profile to an object that contains the required user information from the passport session. This way, I can still associate session data with socket instances without completely reverting to a pre-JWT SocketCluster.

I'm not very happy about having to modify the SC source, but it's the best way I could think of keeping my current workflow intact.

**Comment by jondubois** [comment]
@JasonPaulos The SCSocket object has a request property. So you don't necessarily have to change SC's code. You could use:

```js
socket.request.profile
```

The request object represents the handshake HTTP request which was used to initiate the persistent connection.

This property is not mentioned in the documentation... I will fix that soon.

**Comment by MegaGM** [comment]
SocketCluster definitely should have a bridge ` passport <=> socketcluster `

**Comment by jondubois** [comment]
I'm closing this and opening a separate issue regarding passport - The title of this issue is no longer relevant enough. https://github.com/TopCloud/socketcluster/issues/71

**Comment by jondubois** [comment]
If anyone wants more info on authentication over HTTP, you may want to check out this thread; https://github.com/SocketCluster/socketcluster/issues/233#issuecomment-254871963
